### PR TITLE
PERF: groupby.quantile

### PR DESCRIPTION
**Pull request by jbrockmendel**
Nearly 10x

```
import pandas as pd
import numpy as np

np.random.seed(536445246)
arr = np.random.randn(10**5, 5)
df = pd.DataFrame(arr, columns=["A", "B", "C", "D", "E"])

gb = df.groupby(df.index % 7)

qs = np.arange(0, 1, 0.1)

%timeit res = gb.quantile(qs)
771 ms ± 34 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)  # <- master
81.2 ms ± 8.63 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)  # <- PR
```

### CODE REVIEW

File: pandas/_libs/groupby.pyx [hunk 1]
```diff
@@ -770,25 +770,25 @@ def group_ohlc(floating[:, ::1] out,
 
 @cython.boundscheck(False)
 @cython.wraparound(False)
-def group_quantile(ndarray[float64_t] out,
+def group_quantile(ndarray[float64_t, ndim=2] out,
```
Comment (jreback) [inline pandas/_libs/groupby.pyx:773]: i think we name these _2d (or similar), thinking of rank of these are 2d based?

File: pandas/core/groupby/groupby.py [hunk 2]
```diff
@@ -2523,63 +2535,71 @@ def post_processor(vals: np.ndarray, inference: type | None) -> np.ndarray:
...
+        if is_ser:
+            # i.e. SeriesGroupBy
+            mgr = obj.to_frame()._mgr
+        else:
+            mgr = self._get_data_to_aggregate()
+
+        res_mgr = mgr.grouped_reduce(blk_func, ignore_failures=True)
+        if len(res_mgr.items) != len(mgr.items):
+            warnings.warn(
+                "Dropping invalid columns in "
```
Comment (jreback) [inline pandas/core/groupby/groupby.py:2583]: tests hit this?

File: pandas/core/groupby/groupby.py [hunk 3]
```diff
@@ -2523,63 +2535,71 @@ def post_processor(vals: np.ndarray, inference: type | None) -> np.ndarray:
...
+        qs = np.array(q, dtype=np.float64)
+        ids, _, ngroups = self.grouper.group_info
+        nqs = len(qs)
+
+        func = partial(
+            libgroupby.group_quantile, labels=ids, qs=qs, interpolation=interpolation
+        )
+
+        def blk_func(values: ArrayLike) -> ArrayLike:
+            mask = isna(values)
```
Comment (jreback) [inline pandas/core/groupby/groupby.py:2551]: 0-dim dont' enter here right?

File: pandas/core/groupby/groupby.py [hunk 4]
```diff
@@ -3517,3 +3545,22 @@ def get_groupby(
         mutated=mutated,
         dropna=dropna,
     )
+
+
+def _insert_quantile_level(idx: Index, qs: npt.NDArray[np.float64]) -> MultiIndex:
+    """
+    Insert the sequence 'qs' of quantiles as the inner-most level of a MultiIndex.
+
```
Comment (jreback) [inline pandas/core/groupby/groupby.py:3553]: can you add Paramters / Returns

**Review by jreback — CHANGES_REQUESTED** [review]
looks good!

cc @mzeitlin11

**Comment by jreback** [comment]
prob worth a whatsnew note as large change in perf

File: pandas/_libs/groupby.pyx (hunk 1 above)
Comment (jbrockmendel) [reply pandas/_libs/groupby.pyx:773]: not in this file

File: pandas/core/groupby/groupby.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/core/groupby/groupby.py:2583]: yep

File: pandas/core/groupby/groupby.py (hunk 3 above)
Comment (jbrockmendel) [reply pandas/core/groupby/groupby.py:2551]: correct

File: pandas/core/groupby/groupby.py (hunk 4 above)
Comment (jbrockmendel) [reply pandas/core/groupby/groupby.py:3553]: will do

File: pandas/_libs/groupby.pyx (hunk 1 above)
Comment (jreback) [reply pandas/_libs/groupby.pyx:773]: kk, prob want to make a pass over the cython code to unify the names (e.g. append 1d or 2d) where appropriate

**Comment by jbrockmendel** [comment]
whatsnew added + green

**Review by mzeitlin11 — APPROVED** [review]
LGTM! Might make a followup with couple things I noticed while reading through (but not expressly related to the pr, so no reason to block merging)

**Review by jreback — APPROVED** [review]

**Pull request merged**